src/functions.rs

```diff
@@ -2919,6 +2919,84 @@ pub fn initialize_terminated<'a>(functions: &'a FunctionMap<'a>) -> (Vec<String>
     (term_ret_vec, nonterm_ret_vec)
 }
 
+// Helper function to ressolve the "true" name of a function
+// given the FuncCall (the Function we are looking to get the true cil name),
+// FunctionInfo (for the function in which FuncCall is called) and FunctionMap (all FunctionInfos).
+// If the function call needs to be cast this will handle that.  This function will not
+// validate the cast that is done else where.
+// If the function call has a "this" class or no class, it will substitue the class of
+// the caller
+// Lastely if the function is type aliased this function will resolve the alias.
+fn resolve_true_cil_name(
+    call: FuncCall,
+    func_info: &FunctionInfo,
+    function_map: &FunctionMap,
+) -> Result<String, CascadeErrors> {
+    // First if check is to see if we need to cast the call.
+    let original_cil_name = if let Some(cast_class) = call.cast_name {
+        // This should be mostly safe since we have already looked at castables in prevalidate
+        get_cil_name(Some(&cast_class), &call.name)
+    }
+    // If we are calling something with this or None it must be in the same class
+    // so hand place the class name
+    else if call.class_name.as_ref() == Some(&CascadeString::from("this"))
+        || call.class_name.is_none()
+    {
+        if let FunctionClass::Type(class) = func_info.class {
+            get_cil_name(Some(&class.name), &call.name)
+        } else {
+            return Err(CascadeErrors::from(
+                ErrorItem::make_compile_or_internal_error(
+                    "Could not determine class for 'this.' function call",
+                    Some(func_info.declaration_file),
+                    call.get_name_range(),
+                    "Perhaps you meant to place the function in a resource or domain?",
+                ),
+            ));
+        }
+    } else {
+        call.get_cil_name()
+    };
+    // Finally deal with alias
+    if let Some(call_func_info) = function_map.get(&original_cil_name) {
+        return Ok(call_func_info.get_cil_name());
+    }
+
+    Ok(original_cil_name)
+}
+
+// Search through the non terminating functions to find loops
+fn find_recursion_loop(
+    func: &str,
+    function_map: &FunctionMap,
+    terminated_list: &Vec<String>,
+    visited: &mut Vec<String>,
+) -> Result<Vec<String>, CascadeErrors> {
+    visited.push(func.to_string());
+    if let Some(function_info) = function_map.get(func) {
+        let func_calls = get_all_func_calls(function_info.original_body.to_vec());
```

Comment:
This doesn't need to happen as part of this PR, but it looks like we could just convert get_all_func_calls() to take a slice instead of a vec and avoid constructing a vec here.

Reply:
https://github.com/dburgener/cascade/pull/151